File: examples/src/test/scala/docs/guide/ShoppingCartApp.scala
/*
 * Copyright (C) 2020-2022 Lightbend Inc. <https://www.lightbend.com>
 */

//#guideSetup
package docs.guide

import org.apache.pekko
import pekko.actor.typed.ActorSystem
import pekko.actor.typed.scaladsl.Behaviors
import pekko.projection.ProjectionBehavior
import pekko.projection.eventsourced.EventEnvelope
import com.typesafe.config.ConfigFactory
//#guideSetup

//#guideSourceProviderImports
import org.apache.pekko
import pekko.persistence.cassandra.query.scaladsl.CassandraReadJournal
import pekko.persistence.query.Offset
import pekko.projection.eventsourced.scaladsl.EventSourcedProvider
import pekko.projection.scaladsl.SourceProvider
//#guideSourceProviderImports

//#guideProjectionImports
import org.apache.pekko
import pekko.projection.ProjectionId
import pekko.projection.cassandra.scaladsl.CassandraProjection
import pekko.stream.connectors.cassandra.scaladsl.CassandraSessionRegistry
//#guideProjectionImports

//#guideSetup

object ShoppingCartApp extends App {
  val config = ConfigFactory.load("guide-shopping-cart-app.conf")

  ActorSystem(
    Behaviors.setup[String] { context =>
      val system = context.system

      // ...

      // #guideSetup
      // #guideSourceProviderSetup
      val sourceProvider: SourceProvider[Offset, EventEnvelope[ShoppingCartEvents.Event]] =
        EventSourcedProvider
          .eventsByTag[ShoppingCartEvents.Event](
            system,
            readJournalPluginId = CassandraReadJournal.Identifier,
            tag = ShoppingCartTags.Single)
      // #guideSourceProviderSetup

      // #guideProjectionSetup
      implicit val ec = system.executionContext
      val session = CassandraSessionRegistry(system).sessionFor("pekko.projection.cassandra.session-config")
      val repo = new ItemPopularityProjectionRepositoryImpl(session)
      val projection = CassandraProjection.atLeastOnce(
        projectionId = ProjectionId("shopping-carts", ShoppingCartTags.Single),
        sourceProvider,
        handler = () => new ItemPopularityProjectionHandler(ShoppingCartTags.Single, system, repo))

      context.spawn(ProjectionBehavior(projection), projection.projectionId.id)
      // #guideProjectionSetup

      // #guideSetup
      Behaviors.empty
    },
    "ShoppingCartApp",
    config)
}
//#guideSetup

File: examples/src/test/java/jdocs/guide/ShoppingCartApp.java
/*
 * Copyright (C) 2020-2022 Lightbend Inc. <https://www.lightbend.com>
 */

// #guideSetup
package jdocs.guide;

import org.apache.pekko.actor.typed.ActorSystem;
import org.apache.pekko.actor.typed.javadsl.Behaviors;
import org.apache.pekko.projection.ProjectionBehavior;
import org.apache.pekko.projection.eventsourced.EventEnvelope;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
// #guideSetup

// #guideSourceProviderImports
import org.apache.pekko.persistence.cassandra.query.javadsl.CassandraReadJournal;
import org.apache.pekko.persistence.query.Offset;
import org.apache.pekko.projection.eventsourced.javadsl.EventSourcedProvider;
import org.apache.pekko.projection.javadsl.SourceProvider;
// #guideSourceProviderImports

// #guideProjectionImports
import org.apache.pekko.projection.ProjectionId;
import org.apache.pekko.projection.cassandra.javadsl.CassandraProjection;
import org.apache.pekko.projection.javadsl.AtLeastOnceProjection;
import org.apache.pekko.stream.connectors.cassandra.javadsl.CassandraSession;
import org.apache.pekko.stream.connectors.cassandra.javadsl.CassandraSessionRegistry;
// #guideProjectionImports

// #guideSetup

public class ShoppingCartApp {
  public static void main(String[] args) throws Exception {
    Config config = ConfigFactory.load("guide-shopping-cart-app.conf");

    ActorSystem.create(
        Behaviors.setup(
            context -> {
              ActorSystem<Void> system = context.getSystem();

              // ...

              // #guideSetup
              // #guideSourceProviderSetup
              SourceProvider<Offset, EventEnvelope<ShoppingCartEvents.Event>> sourceProvider =
                  EventSourcedProvider.eventsByTag(
                      system, CassandraReadJournal.Identifier(), ShoppingCartTags.SINGLE);
              // #guideSourceProviderSetup

              // #guideProjectionSetup
              CassandraSession session =
                  CassandraSessionRegistry.get(system)
                      .sessionFor("pekko.projection.cassandra.session-config");
              ItemPopularityProjectionRepositoryImpl repo =
                  new ItemPopularityProjectionRepositoryImpl(session);
              AtLeastOnceProjection<Offset, EventEnvelope<ShoppingCartEvents.Event>> projection =
                  CassandraProjection.atLeastOnce(
                      ProjectionId.of("shopping-carts", ShoppingCartTags.SINGLE),
                      sourceProvider,
                      () ->
                          new ItemPopularityProjectionHandler(
                              ShoppingCartTags.SINGLE, system, repo));

              context.spawn(ProjectionBehavior.create(projection), projection.projectionId().id());
              // #guideProjectionSetup

              // #guideSetup
              return Behaviors.empty();
            }),
        "ShoppingCartApp",
        config);
  }
}
// #guideSetup
